packages/orchestration/src/orchestration-chat-completion-stream.ts

```diff
@@ -0,0 +1,170 @@
+import { createLogger } from '@sap-cloud-sdk/util';
+import { SseStream } from '@sap-ai-sdk/core';
+import { OrchestrationChatCompletionStreamChunkResponse } from './orchestration-chat-completion-stream-chunk-response.js';
+import type { CompletionPostResponseStreaming } from './client/api/schema/index.js';
+import type { HttpResponse } from '@sap-cloud-sdk/http-client';
+import type { OrchestrationChatCompletionStreamResponse } from './orchestration-chat-completion-stream-response.js';
+
+const logger = createLogger({
+  package: 'orchestration',
+  messageContext: 'orchestration-chat-completion-stream'
+});
+
+/**
+ * Chat completion stream containing post-processing functions.
+ */
+export class OrchestrationChatCompletionStream<Item> extends SseStream<Item> {
+  /**
+   * Create a chat completion stream based on the http response.
+   * @param response - Http response.
+   * @returns Chat completion stream.
+   * @internal
+   */
+  public static _create(
+    response: HttpResponse,
+    controller: AbortController
+  ): OrchestrationChatCompletionStream<CompletionPostResponseStreaming> {
+    const stream = SseStream.transformToSseStream<any>(response, controller); // TODO: Check if this can be narrowed
+    return new OrchestrationChatCompletionStream(stream.iterator, controller);
+  }
+
+  /**
+   * Wrap raw chunk data with chunk response class to provide helper functions.
+   * @param stream - Chat completion stream.
+   * @internal
+   */
+  static async *_processChunk(
+    stream: OrchestrationChatCompletionStream<CompletionPostResponseStreaming>
+  ): AsyncGenerator<OrchestrationChatCompletionStreamChunkResponse> {
+    for await (const chunk of stream) {
+      yield new OrchestrationChatCompletionStreamChunkResponse(chunk);
+    }
+  }
+
+  /**
+   * @internal
+   */
+  static async *_processFinishReason(
+    stream: OrchestrationChatCompletionStream<OrchestrationChatCompletionStreamChunkResponse>,
+    response?: OrchestrationChatCompletionStreamResponse<CompletionPostResponseStreaming>
+  ): AsyncGenerator<OrchestrationChatCompletionStreamChunkResponse> {
+    for await (const chunk of stream) {
+      chunk.data.orchestration_result?.choices.forEach((choice: any) => {
+        const choiceIndex = choice.index;
+        if (choiceIndex >= 0) {
+          const finishReason = chunk.getFinishReason(choiceIndex);
+          if (finishReason) {
+            if (response) {
+              response._getFinishReasons().set(choiceIndex, finishReason);
+            }
+            switch (finishReason) { // TODO: Cover all finish reasons
```

Comment:
[q] Is there any other finish reason? Can you share this knowledge with me? Pretty curious.

Reply:
I was under the assumption that other filters also write into the finish reason, but apparently they throw 400 instead.
I've removed the todo.